src/d3a/d3a_core/sim_results/file_export_endpoints.py

```diff
@@ -344,76 +342,83 @@ def _specific_row(self, slot, market):
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                    self.total_self_consumption_wh += trade.offer.energy * 1000
                 else:
-                    if trade.buyer is child.name:
-                        self._total_energy += trade.offer.energy
-
-    def _export_house_pv_self_consumption(self, area):
-        house_pv_device = list()
-        house_load_device = list()
-        trade_by_pv = 0
-        total_energy_bought = 0
-        traded_from_pv = 0
-        if not _is_house_node(area):
-            return
-        for child in area.children:
-            if isinstance(child.strategy, PVStrategy):
-                house_pv_device.append(child.name)
-            elif _is_load_node(child) or _is_prosumer_node(child):
-                house_load_device.append(child.name)
-
-        for markets in area.past_markets:
-            for trade in markets.trades:
-                # Total Electricity traded by house device
-                if trade.buyer in house_load_device:
-                    total_energy_bought += trade.offer.energy
-                # Electricity produced by PV
-                if trade.offer.seller in house_pv_device:
-                    trade_by_pv += trade.offer.energy
-                # PV electricity self_consumption
-                if trade.offer.seller in house_pv_device and trade.buyer in house_load_device:
-                    traded_from_pv += trade.offer.energy
-
-        if trade_by_pv != 0:
-            self_consumption_within_house = traded_from_pv / trade_by_pv
-        else:
-            self_consumption_within_house = 0
-        if total_energy_bought != 0:
-            self_sufficiency = traded_from_pv / total_energy_bought
-        else:
-            self_sufficiency = 0
-
-        return {"self_consumption_within_house": self_consumption_within_house,
-                "self_sufficiency": self_sufficiency}
+                    self.total_self_consumption_wh += self.self_consumption_buffer_wh
+                    self.self_consumption_buffer_wh = 0
+            elif (trade.buyer_origin not in self.consumer_list or
+                  trade.buyer_origin not in self.ess_list) and self.self_consumption_buffer_wh > 0:
+                if (self.self_consumption_buffer_wh - trade.offer.energy * 1000) > 0:
+                    self.self_consumption_buffer_wh -= trade.offer.energy * 1000
+                else:
+                    self.self_consumption_buffer_wh = 0
+
+    def _accumulate_energy_trace(self):
+        for c_area in self.consumer_area_list:
+            for market in c_area.past_markets:
+                for trade in market.trades:
+                    self._accumulate_self_consumption(trade)
+                    self._accumulate_self_consumption_buffer(trade)
+                    self._dissipate_self_consumption_buffer(trade)
+
+    def _area_self_sufficiency(self, area):
+        self.producer_list = list()
```

Comment:
Well, I am not sure whether it is that safe to create member variables outside of the constructor. We can discuss also offline about the potential structure, but I would feel much better if all self-sufficiency is encapsulated in a class. The reason I am asking this is because there are going to be more KPIs quite soon and I think there could be conflicts between these variables, which could be avoided if all these were inside a class.